### DISCUSSION TITLE
Can someone provide better contains() usage documentation?

### DISCUSSION THREAD

**Discussion by rromanchuk** [realm/realm-cocoa#7610] [Q&A]
> Equivalent to the IN operator. Evaluates to true if the property referenced by the expression contains the value.

I don't understand how the example is showing, what i assumed would be a typed equivalent to `.filter("assignee IN %@", ["Ali", "Jamie"])`

```swift
let aliOrJamiesTasks = tasks.where {
    $0.assignee.contains("Ali") || $0.assignee.contains("Jamie")
}
print("Tasks IN Ali or Jamie: \(aliOrJamiesTasks.count)")
```
This just seems like a much more annoying way to write `.where { $0.assignee == "Ali" ||  $0.assignee == "Jamie" }`

For example, how would you compose `.filter("challenge.id == %@ && (status IN %@ || gave_up == true)", challengeId, [PlayerStatus.kicked_out.rawValue, PlayerStatus.left.rawValue])`

```swift
class Player: Object, JapxDecodable, WithSaveHooks {
  enum PlayerStatus: String, Codable, PersistableEnum, _Indexable {
    case joined
    case left
    case kicked_out
    case loser
    case winner
    case unknown
  }
  
  @Persisted(primaryKey: true) var id = ""
  @Persisted(indexed: true) var status = PlayerStatus.joined
  @Persisted var challenge: Challenge?
}
```

I would assume the block would look something like this`.where { [.left, .kicked_out].contains($0.status) }`  If anyone has more realistic examples to get me going, before I head to the reference docs, which is dense with generics, i'll give you a 🥇

**Comment by leemaguire** [comment]
I think what you want is `containsAny(in:)`. So you would do something like:
```swift
.where { $0.assignee.containsAny(in: ["Ali", "Jamie"]) }
```

This produces the NSPredicate string `"(ANY assignee IN %@)"`.

**Reply by rhys-rant** [reply]
I'm having a similar problem to this.
My equivalent of
```swift
realm.objects(Player.self).where { player in
  player.status
```

code doesn't give me `.containsAny(in:)` as an option.

I only get `.contains(range: (Closed)Range<Player.PlayerStatus>)` as options. I get `No exact matches in reference to subscript` as the error.

**Comment by leemaguire** [comment]
@rhys-rantmedia `containsAny(in:)` is for collection properties. Can you show an example of what you are trying to achieve?

**Reply by rhys-rant** [reply]
I thought as much.
So say I have the following types:

```swift
class Order: Object {
  @Persisted var name: String
  @Persisted var date: Date = Date()
  @Persisted var status: Status
}

enum Status: Int, PersistableEnum {
  case inProgress
  case delivered
  case cancelled
}
```

With the `NSPredicate` style, if I want to get all `Order` objects which are either `.delivered` or `.cancelled`, I would write something like this:

```swift
realm.objects(Order.self)
  .filter("status IN %@", [Status.delivered.rawValue, Status.cancelled.rawValue])
```

How do I do the equivalent with the type-safe `Query`, if it's currently even possible? As @rromanchuk said, it seems like the logical syntax would be something like:

```swift
realm.objects(Order.self)
  .where { order in [.delivered, .cancelled].contains(order.status) }
```

or something along those lines. But that's not possible, because `order.status` is type `Query<Status>`.

Is there a method on `Query` that I'm missing so I *can* write something like
```swift
realm.objects(Order.self)
  .where { order in order.status.in([.delivered, .cancelled]) }
```

**Reply by leemaguire** [reply]
That is not a valid use of `IN` in this scenario. `IN` should be used with collection types. I suggest:
```swift
realm.objects(Order.self)
            .where { order in order.status == .delivered || order.status == .cancelled }
```

**Reply by rhys-rant** [reply]
> That is not a valid use of `IN` in this scenario. `IN` should be used with collection types. I suggest:
> 
> ```swift
> realm.objects(Order.self)
>             .where { order in order.status == .delivered || order.status == .cancelled }
> ```

I guess that's the problem - the `NSPredicate` `IN` query works fine. I can check if a single value on the left-hand side is `IN` an array/list/set that's the right-hand side using an `NSPredicate`.

The more real-world example is I have a method that is currently structured like this:

```swift
func allOrders(statuses: [Status] = [], searchTerms: String) -> Results<Status> {
  var results = realm.objects(Order.self)
    .filter("status IN %@", statuses)

  // More filtering here, using searchTerms

  return results
}
```

and have done similar in other applications without issue. I'd like to be able to pass in, at `allOrders` call-time, the statuses I'm filtering by, given other (non-trivial) work happens *after* that initial filter takes place. That's why I can't just use multiple `||` statements inside the `where`.

Ideally, I'd like to be able to do something more like `statuses.contains(order.status)` (where collection is the non-Query-wrapped type), or have a method on `Query` to check against multiple possible values, generating `NSPredicate` string `"status IN {2, 3}"`

The same behaviour - `value IN collection` - works in older Cocoa APIs as expected FWIW, as demo'd here: https://gist.github.com/rhys-rantmedia/20250d8bd79a7e0cff67aaf970490980 so I _think_ my use of `IN` is appropriate, at least with regards to how Foundation handles NSPredicates.

Either this just isn't possible right now, and I guess this is a feature request, or there's just something I'm missing 😄

**Reply by rromanchuk** [reply]
Didn't realize there were replies to this.  @rhys-rantmedia same, i have quite a bit of `NSPredicate`  usage  using `status IN %@` too. Did you poke around this issue any more?
